In [1]:
# Importing Packages
import pandas as pd
import numpy as np

In [2]:
# Loading Comma Seperated Data using read_table pandas function in 'spam_train' dataframe
spam_train = pd.read_table("spambasetrain.csv", sep=",", header=None)

# Adding Headers to Data
spam_train.columns = ["char_freq_;", "char_freq_(", "char_freq_[", "char_freq_!", "char_freq_$", "char_freq_#", "capital_run_length_average", "capital_run_length_longest", "capital_run_length_total", "spam_label"]
spam_train.to_csv("spambasetrain_wheader.csv")
spam_train

char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  \
0           0.000        0.316        0.000        0.000        0.000   
1           0.000        0.735        0.000        0.735        0.000   
2           0.000        0.105        0.000        0.000        0.158   
3           0.000        0.299        0.000        0.000        0.000   
4           0.000        0.000        0.000        0.817        0.000   
5           0.000        0.000        0.000        2.272        0.000   
6           0.055        0.334        0.000        0.055        0.000   
7           0.000        0.162        0.000        0.000        0.000   
8           0.000        0.075        0.000        0.613        0.532   
9           0.000        0.052        0.000        0.078        0.000   
10          0.000        0.204        0.000        0.204        0.127   
11          0.000        0.000        0.000        0.000        0.000   
12          0.000        0.000        0.119        0.000        0.000   
13          0.000        0.000        0.000        0.000        0.000   
14          0.000        0.000        0.000        0.166        0.000   
15          0.000        0.099        0.000        0.396        0.099   
16          0.000        0.413        0.000        0.000        0.000   
17          0.000        0.310        0.000        0.517        0.000   
18          0.000        0.000        0.000        0.396        0.000   
19          0.000        0.077        0.000        0.000        0.000   
20          0.000        0.000        0.000        0.101        0.304   
21          0.000        0.529        0.000        0.000        0.000   
22          0.000        0.211        0.000        0.000        0.211   
23          0.000        0.281        0.000        0.210        0.000   
24          0.000        0.000        0.000        0.165        0.000   
25          0.000        0.362        0.000        0.000        0.000   
26          0.029        0.119        0.119        0.000        0.000   
27          0.000        0.255        0.102        0.000        0.000   
28          0.000        0.081        0.000        0.486        0.243   
29          0.561        0.140        0.000        0.000        0.000   
...           ...          ...          ...          ...          ...   
2969        0.000        0.156        0.000        0.312        0.000   
2970        0.000        0.000        0.000        0.000        0.000   
2971        0.000        0.684        0.000        0.000        0.000   
2972        0.000        0.276        0.000        0.552        0.000   
2973        0.000        0.000        0.000        0.000        0.000   
2974        0.010        0.052        0.000        0.010        0.169   
2975        0.011        0.041        0.071        0.379        0.136   
2976        0.000        0.000        0.000        0.000        0.000   
2977        0.000        0.000        0.000        0.368        0.552   
2978        0.000        0.149        0.000        0.149        0.074   
2979        0.000        0.000        0.000        0.578        0.000   
2980        0.000        0.000        0.000        0.274        0.000   
2981        0.000        0.116        0.000        0.232        0.000   
2982        0.000        0.049        0.000        0.049        0.000   
2983        0.000        0.168        0.000        0.112        0.000   
2984        0.000        0.000        0.000        0.000        0.000   
2985        0.000        0.189        0.000        0.441        0.000   
2986        0.000        0.065        0.000        0.318        0.754   
2987        0.000        0.045        0.000        0.000        0.091   
2988        0.000        0.000        0.000        0.117        0.117   
2989        0.000        0.000        0.000        0.574        0.000   
2990        0.000        0.300        0.000        0.000        0.000   
2991        0.000        0.123        0.000        0.197        

# Training

In [3]:
# Counting Total Labels
total_labels = spam_train["spam_label"].count()

print("Total Label Count:", total_labels)

Total Label Count: 2999


In [4]:
# Count Per Label
value_counts = spam_train["spam_label"].value_counts()
class0_count = value_counts[0]
class1_count = value_counts[1]

print("Class 0 Label Count:", class0_count)
print("Class 1 Label Count:", class1_count)

Class 0 Label Count: 1794
Class 1 Label Count: 1205


## Estimation of P(C): 
To estimate the P(C) values, just calculate the fraction of the training examples that are in class C. For example, to estimate P(C = 1) calculate:

\begin{equation*}
\frac{number of examplesin training file in Class 1}{total number of examples in training file}
\end{equation*}

In [5]:
# Estimation of P(C)
prob_spam = class1_count/total_labels
prob_not_spam = class0_count/total_labels

## 1. What was the estimated value of P(C) for C = 1?

In [6]:
print("P(Spam)=", prob_spam)

P(Spam)= 0.4018006002


## 2. What was the estimated value of P(C) for C = 0?

In [7]:
print("P(Not Spam)=", prob_not_spam)

P(Not Spam)= 0.5981993998


In [8]:
# Seperating Data According to Class Labels
class0 = spam_train[spam_train['spam_label']==0]
class1 = spam_train[spam_train['spam_label']==1]
del class0['spam_label']
del class1['spam_label']

## Estimation of parameters of Gaussian pdf

In [9]:
# Calculation of Class 0 Mean
class0_mean = class0.describe().T['mean']
print("Class 0 Mean Calculation:")
class0_mean

Class 0 Mean Calculation:


char_freq_;                     0.048426
char_freq_(                     0.157487
char_freq_[                     0.016960
char_freq_!                     0.110713
char_freq_$                     0.011503
char_freq_#                     0.025892
capital_run_length_average      2.376590
capital_run_length_longest     18.182274
capital_run_length_total      164.143255
Name: mean, dtype: float64

In [10]:
# Calculation of Class 1 Mean
class1_mean = class1.describe().T['mean']
print("Class 1 Mean Calculation:")
class1_mean

Class 1 Mean Calculation:


char_freq_;                     0.020247
char_freq_(                     0.103439
char_freq_[                     0.007847
char_freq_!                     0.491662
char_freq_$                     0.173151
char_freq_#                     0.084901
capital_run_length_average      9.496347
capital_run_length_longest     97.209129
capital_run_length_total      453.512863
Name: mean, dtype: float64

In [11]:
# Calculation of Variance according to the given formula in assignment
def cal_variance(column, mean, count):
    return (column.apply(lambda x: (x-mean)**2).sum())/(count-1)

class0_variance_list = []
class1_variance_list = []

for column in class0:
    class0_variance_list.append(cal_variance(class0[column], class0_mean[column], class0_count))
for column in class1:
    class1_variance_list.append(cal_variance(class1[column], class1_mean[column], class1_count))

In [12]:
# Calculation of Class 0 Variance
class0_variance = pd.Series(class0_variance_list, index=class0.columns)
print("Class 0 Variance Calculation:")
class0_variance

Class 0 Variance Calculation:


char_freq_;                        0.088306
char_freq_(                        0.070180
char_freq_[                        0.004845
char_freq_!                        0.417588
char_freq_$                        0.004080
char_freq_#                        0.089269
capital_run_length_average        37.447271
capital_run_length_longest      1915.266813
capital_run_length_total      144134.168535
dtype: float64

In [13]:
# Calculation of Class 1 Variance
class1_variance = pd.Series(class1_variance_list, index=class1.columns)
print("Class 1 Variance Calculation:")
class1_variance

Class 1 Variance Calculation:


char_freq_;                        0.007542
char_freq_(                        0.032143
char_freq_[                        0.002416
char_freq_!                        0.547792
char_freq_$                        0.129976
char_freq_#                        0.527283
capital_run_length_average      2965.984486
capital_run_length_longest     36369.991113
capital_run_length_total      614182.120474
dtype: float64

## 3. What were the estimated values for (mean, variance) for the Gaussian corresponding to attribute capital run length longest and class 1 (Spam).

In [14]:
print("Mean for capital run length longest and class 1=",class1_mean['capital_run_length_longest'])
print("Variance for capital run length longest and class 1=",class1_variance['capital_run_length_longest'])

Mean for capital run length longest and class 1= 97.2091286307
Variance for capital run length longest and class 1= 36369.9911126


## 4. What were the estimated values for (mean, variance) for the Gaussian corresponding to attribute char freq ; and Class 0.

In [15]:
print("Mean for char_freq_; and class 0=",class0_mean['char_freq_;'])
print("Variance for char_freq_; and class 0=",class0_variance['char_freq_;'])

Mean for char_freq_; and class 0= 0.0484258639911
Variance for char_freq_; and class 0= 0.0883056032571


In [16]:
# Gaussian PDF Formula
import math
def normpdf(x, mean, var):
    pi = 3.1415926
    denom = (2*pi*var)**.5
    num = math.exp(-(float(x)-float(mean))**2/(2*var))
    return num/denom

# Testing

In [17]:
# Loading Comma Seperated Data using read_table pandas function in 'spam_test' dataframe
spam_test = pd.read_table("spambasetest.csv", sep=",", header=None)

# Adding Headers to Data
spam_test.columns = ["char_freq_;", "char_freq_(", "char_freq_[", "char_freq_!", "char_freq_$", "char_freq_#", "capital_run_length_average", "capital_run_length_longest", "capital_run_length_total", "spam_label"]
spam_test

char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  \
0          0.000        0.000        0.000        0.000        0.000   
1          0.000        0.000        0.000        0.102        0.000   
2          0.000        0.492        0.000        0.000        0.000   
3          0.203        0.195        0.050        0.000        0.014   
4          0.000        0.000        0.000        0.874        0.000   
5          0.000        0.793        0.000        0.000        0.000   
6          0.000        0.000        0.000        0.170        0.170   
7          0.000        0.132        0.000        0.022        0.000   
8          0.000        0.000        0.000        0.000        0.000   
9          0.000        0.000        0.000        0.277        0.000   
10         0.000        0.000        0.000        0.079        0.106   
11         0.000        0.203        0.000        0.000        0.000   
12         0.000        0.088        0.000        0.000        0.000   
13         0.000        0.000        0.000        0.000        0.000   
14         0.000        0.000        0.000        0.000        0.000   
15         0.127        0.095        0.000        0.000        0.031   
16         0.123        0.309        0.030        0.000        0.000   
17         0.000        0.295        0.000        0.000        0.000   
18         0.317        0.000        0.000        0.952        0.000   
19         0.000        0.059        0.000        0.434        0.197   
20         0.000        0.000        0.000        0.855        0.794   
21         0.044        0.222        0.000        0.178        0.000   
22         0.000        0.137        0.000        0.045        0.045   
23         0.000        0.257        0.000        0.600        0.429   
24         0.000        0.687        0.000        0.000        0.000   
25         0.714        0.000        0.000        0.238        0.000   
26         0.000        0.452        0.000        2.865        0.301   
27         0.000        0.083        0.000        0.503        0.000   
28         0.000        0.000        0.000        0.000        0.000   
29         0.000        0.318        0.000        0.000        0.000   
..           ...          ...          ...          ...          ...   
170        0.000        0.526        0.000        0.000        0.000   
171        0.018        0.073        0.000        0.000        0.000   
172        0.153        0.092        0.000        0.000        0.000   
173        0.134        0.089        0.000        0.000        0.000   
174        0.000        0.274        0.000        0.000        0.000   
175        0.000        0.000        0.000        0.467        0.233   
176        0.000        0.000        0.000        0.427        0.000   
177        0.000        0.075        0.000        0.131        0.056   
178        0.000        0.000        0.000        1.282        0.000   
179        0.000        0.037        0.037        1.362        0.037   
180        0.015        0.180        0.000        0.045        0.015   
181        0.007        0.304        0.000        0.053        0.030   
182        0.182        0.182        0.000        0.000        0.000   
183        0.000        0.000        0.000        0.000        0.000   
184        0.000        0.054        0.000        0.000        0.271   
185        0.000        0.165        0.000        0.371        0.061   
186        0.000        0.105        0.000        0.000        0.000   
187        0.000        0.022        0.000        0.000        0.000   
188        0.085        0.198        0.056        0.000        0.000   
189        0.000        0.170        0.000        0.510        0.340   
190        0.000        0.077        0.064        0.348        0.103   
191        0.000        0.000        0.256        0.128        0.000   
192        0.000        0.000        0.000        0.000        0.000   
193        0.000        0.048        0.000   

In [18]:
# Predictions for Test Dataset
import math
import xlwt

workbook = xlwt.Workbook()  
sheet = workbook.add_sheet("programanswers")

i = 0
correct_count =0
answer_list = []

if prob_not_spam > prob_spam:
    zero_r_label = 0
else:
    zero_r_label = 1

c_label =0    
for index, row in spam_test.iterrows():
      # Taking Natural Log for P(C)
      pr0 =  math.log(prob_not_spam) 
      pr1 =  math.log(prob_spam)
     
      for column in spam_test:
           # Not counting "spam_label" while doing Predictions!
           if (column!="spam_label"):
               #print(row[column])
               pr0 += math.log(normpdf(row[column],class0_mean[column],class0_variance[column]))
               #print(pr0)
               pr1 += math.log(normpdf(row[column],class1_mean[column],class1_variance[column]))
               #print(pr1)
    
      print("P(X|C=0) = ",pr0)  
      print("P(X|C=1) = ",pr1)
    
      # Label the example with the class achieving the maximum value for this expression. If there is a tie, give the example the label 1.
      if (pr0 > pr1):
           class_label = 0
      else:
           class_label = 1 
      print("class_label is : ",class_label)
      answer_list.append(class_label)
      #print(class_label)
      
      if (row["spam_label"] == zero_r_label):
        c_label += 1
    
      # Comparing results with actual predictions
      if  row["spam_label"] == class_label:
        result= "Match"
        correct_count = correct_count+1
      else:
        result="No match"
      sheet.write(i, 0,row["spam_label"] ) 
      sheet.write(i, 1,class_label) 
      sheet.write(i, 2,result) 
      i=i+1
        
         
workbook.save("output.xls")
print("\nNo. of Correct Predictions = ", correct_count)

P(X|C=0) =  -11.107084331454885
P(X|C=1) =  -17.119278957199707
class_label is :  0
P(X|C=0) =  -11.012097539615262
P(X|C=1) =  -16.92852455575198
class_label is :  0
P(X|C=0) =  -11.605452996414062
P(X|C=1) =  -19.202399754584867
class_label is :  0
P(X|C=0) =  -14.08967563178135
P(X|C=1) =  -19.67965875296875
class_label is :  0
P(X|C=0) =  -13.809050119823386
P(X|C=1) =  -16.80602092429702
class_label is :  0
P(X|C=0) =  -13.740129786985928
P(X|C=1) =  -24.305689971175273
class_label is :  0
P(X|C=0) =  -14.372713740246251
P(X|C=1) =  -16.73624800106599
class_label is :  0
P(X|C=0) =  -11.72101383805511
P(X|C=1) =  -16.673235030291153
class_label is :  0
P(X|C=0) =  -11.041746286081368
P(X|C=1) =  -17.07838743429091
class_label is :  0
P(X|C=0) =  -11.034318993466666
P(X|C=1) =  -16.876204842153328
class_label is :  0
P(X|C=0) =  -12.114504750694381
P(X|C=1) =  -16.756681312976575
class_label is :  0
P(X|C=0) =  -10.913798633871322
P(X|C=1) =  -17.083732698634456
class_label is :  0

In [19]:
# Predictions from Created Modelling Library
answer_list = np.array(answer_list)
print(answer_list)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 1 1 0 0 1 0 0 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0
 0 0 0 1 1 0 0 0 0 1 0 0 0 1 0 1 1 0 0 0 0 0 1 0 0 0 1 1 1 1 0 1 0 0 0 0 0
 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 0 0 1 0 1 0 0 0 0 0 0 0 0 1 1 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1
 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0]


## 5. Which classes were predicted for the first 5 examples in the test set?

In [20]:
print(answer_list[0:5])

[0 0 0 0 0]


## 6. Which classes were predicted for the last 5 examples in the test set?

In [21]:
print(answer_list[195:200])

[0 0 0 0 0]


## 7. What was the percentage error on the examples in the test file?

In [22]:
total_test_labels = spam_test['spam_label'].count()
print("\nNo. of Incorrect Predictions = ", total_test_labels - correct_count)
percentage_error = (1 - (correct_count/total_test_labels))*100
print("\nPrecentage Error = ",percentage_error)


No. of Incorrect Predictions =  40

Precentage Error =  20.0


# Using Zero-R Classifier Methodology

## 8. Performance Evaluation using Zero-R Classifier

In [23]:
print("\nzero_r_label = ",zero_r_label)
print("\nMaximized Class Correct Predictions = ",c_label)
zero_r_accuracy = c_label/total_test_labels
print ("\nZero-R Classifier Accuracy = ",zero_r_accuracy )


zero_r_label =  0

Maximized Class Correct Predictions =  118

Zero-R Classifier Accuracy =  0.59


The accuracy of Zero-R Classifier is ~59% which is low as compared to performance of Naive-Bayes Classifier.

## 9. Are assumptions reasonable for a spam dataset?

### A Naive Bayes classifier assumes that the presence of a particular feature in a class is unrelated to the presence of any other feature. 

1. Consider a scenario where we have to identify whether a fruit is an apple or not.  
Some parameters may be considered to be an apple if it is red, round, and about 3 inches in diameter. Even if these features depend on each other or upon the existence of the other features, all of these properties independently contribute to the probability that this fruit is an apple and that is why it is known as ‘Naive’.

2. Similarly, we can consider this scenario applying to our prediction whether the email is "Spam" or "Not Spam".  
Features like "char_freq_;", "char_freq_(", "char_freq_[", "char_freq_!", "char_freq_$", "char_freq_#" are independent of each other as each field represents count of different characters respectively.  
Features such as "capital_run_length_average", "capital_run_length_longest", "capital_run_length_total" are also independent from each other as they are completely dependent on changes in input data(word).  
Also, when assumption of independence holds, a Naive Bayes classifier performs better compare to other models like Zero-R Classifier.  
Therefore, we can say that the existence of other features are not dependent on each other and we can classify this problem using Naive Bayes.
